scipy/optimize/_constraints.py

```diff
@@ -283,3 +285,166 @@ def strict_bounds(lb, ub, keep_feasible, n_vars):
+                    dy_all = dy_all.todense()
+                dy_all = np.atleast_2d(dy_all)
+                dy[:n_bound_below, :] = dy_all[i_bound_below]
+                dy[n_bound_below:, :] = -dy_all[i_bound_above]
+                return dy
+            cineq[0]["jac"] = j_ineq
+
+    old_constraints = ceq + cineq
+
+    if len(old_constraints) > 1:
+        warn("Equality and inequality constraints are specified in the same "
+             "element of the constraint list. For efficient use with this "
+             "method, equality and inequality constraints should be specified "
+             "in separate elements of the constraint list. ", OptimizeWarning)
+    return old_constraints
+
+
+def old_constraint_to_new(ic, con):
+
+    # check type
+    try:
+        ctype = con['type'].lower()
+    except KeyError:
+        raise KeyError('Constraint %d has no type defined.' % ic)
+    except TypeError:
+        raise TypeError('Constraints must be a sequence of dictionaries.')
+    except AttributeError:
+        raise TypeError("Constraint's type must be a string.")
+    else:
+        if ctype not in ['eq', 'ineq']:
+            raise ValueError("Unknown constraint type '%s'." % con['type'])
+    if 'fun' not in con:
+        raise ValueError('Constraint %d has no function defined.' % ic)
+
+    lb = 0
+    if ctype == 'eq':
+        ub = 0
+    else:
+        ub = np.inf
+
+    jac = '2-point'
+    if 'args' in con:
+        args = con['args']
+        fun = lambda x: con['fun'](x, *args)
+        if 'jac' in con:
+            jac = lambda x: con['jac'](x, *args)
+    else:
+        fun = con['fun']
+        if 'jac' in con:
+            jac = con['jac']
+
+    return NonlinearConstraint(fun, lb, ub, jac)
+
+
+def standardize_bounds(bounds, x0, meth):
+    if meth == 'trust-constr':
+        if not isinstance(bounds, Bounds):
+            lb, ub = old_bound_to_new(bounds)
+            bounds = Bounds(lb, ub)
+    elif meth in ('l-bfgs-b', 'tnc', 'slsqp'):
+        if isinstance(bounds, Bounds):
+            bounds = new_bounds_to_old(bounds.lb, bounds.ub, x0.shape[0])
+    return bounds
+
+
+def standardize_constraints(constraints, x0, meth):
+    all_constraint_types = (NonlinearConstraint, LinearConstraint, dict)
+    new_constraint_types = all_constraint_types[:-1]
+    if isinstance(constraints, all_constraint_types):
+        constraints = [constraints]
+    constraints = list(constraints)
+
+    additional_constraints = []
+    for i, con in enumerate(constraints):
+        constraints = list(constraints)
+        if (meth == 'trust-constr' and not
+                isinstance(con, new_constraint_types)):
+            constraints[i] = old_constraint_to_new(i, con)
+        elif (meth != 'trust-constr' and
+              isinstance(con, new_constraint_types)):
+            old_constraints = new_constraint_to_old(con, x0)
+            constraints[i] = old_constraints[0]
+            additional_constraints = old_constraints[1:]
+
```

Comment:
@pv agree with `meth==` on top level so that the branching only has to happen once.

But I don't understand how your code `constraints.extend(old_constraints[1:])` agrees with your comment "...adding new items should be avoided". I had the `additional_constraints` list that gets extended (via `+=`) only after the loop is done because I wanted to avoid adding new items during the loop.

Which would you prefer?

(in any case, thanks for pointing it out, because it looks like there was a bug in my code here. `additional_constraints = old_constraints[1:]` should have been `additional_constraints.extend(old_constraints[1:])`)

Reply:
Also, why did you write `for i, con in enumerate(list(constraints)):` rather than `constraints = list(constraints)`? I wanted to ensure that `constraints` is a (mutable) list as I imagine the user could have passed in some other sort of sequence of constraints.

Reply:
It's somewhat simpler to iterate over a copy of the list and mutate the original,
rather than iterating over the original and mutating the copy.

Reply:
I know they're different.  
That approach makes sense.
I didn't see before that we were using `list (constraints)` in different ways. I was using it to ensure that it is mutable. You were using it to create a copy to iterate over. I think we actually need both.
Thanks!